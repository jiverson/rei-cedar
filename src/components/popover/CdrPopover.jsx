import tabbable from 'tabbable';
import style from './styles/CdrPopover.scss';
import propValidator from '../../utils/propValidator';
import IconXSm from '../icon/comps/x-sm';
import CdrButton from '../button/CdrButton';
import CdrPopup from '../popup/CdrPopup';

export default {
  name: 'CdrPopover',
  components: {
    CdrPopup,
    CdrButton,
    IconXSm,
  },
  props: {
    position: {
      type: String,
      required: false,
      default: 'up',
      validator: (value) => propValidator(
        value,
        ['up', 'down', 'left', 'right'],
      ),
    },
    autoPosition: {
      type: Boolean,
      required: false,
      default: true,
    },
    label: {
      type: String,
      required: false,
    },
  },
  data() {
    return {
      style,
      open: false,
      openHandler: undefined,
      lastActive: undefined,
    };
  },
  mounted() {
    this.addHandlers();
  },
  methods: {
    openPopover() {
      const { activeElement } = document;

      this.lastActive = activeElement;
      this.open = true;
      this.$nextTick(() => {
        const tabbables = tabbable(this.$refs.popup.$el);
        if (tabbables[0]) tabbables[0].focus();
      });
    },
    closePopover() {
      this.open = false;
      if (this.lastActive) this.lastActive.focus();
    },
    addHandlers() {
      this.openHandler = this.openPopover.bind(this);
      const triggerElement = this.$refs.trigger.children[0];
      if (triggerElement) {
        triggerElement.addEventListener('click', this.openHandler);
      }
    },
  },
  render() {
    return (
      <div class={this.style['cdr-popover--wrapper']}>
        <div ref="trigger">
          { this.$slots.trigger }
        </div>
        <cdr-popup
          class={this.style['cdr-popover']}
          role="dialog"
          ref="popup"
          position={ this.position }
          autoPosition={ this.autoPosition }
          opened={ this.open }
          onClosed={ this.closePopover }
        >
          <div class={this.style['cdr-popover__container']}>
            <div class={this.style['cdr-popover__content']}>
              {
                (this.$slots.title || this.label) && (
                  <div class={this.style['cdr-popover__title']}>
                    {
                      this.$slots.title
                    }
                    {
                      !this.$slots.title && this.label && (
                        <span>
                          {this.label}
                        </span>
                      )
                    }
                  </div>
                )
              }
              { this.$slots.default }
            </div>
            <cdr-button
              class={this.style['cdr-popover__close-button']}
              icon-only
              on-click={this.closePopover}
              aria-label="Close"
              size="small"
            >
              <icon-x-sm
                slot="icon"
                inherit-color
              />
            </cdr-button>
          </div>
        </cdr-popup>
      </div>
    );
  },
};
