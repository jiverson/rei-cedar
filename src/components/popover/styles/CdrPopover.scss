@import '../../../css/settings/index.scss';

@mixin arrow($borderSide) {
  &::before,
  &::after {
    border-style: solid;
    border-color: transparent;
    #{$borderSide}: 100%;
  }

  &::before {
    border-#{$borderSide}-color: $cdr-color-border-secondary;
    border-width: 10px;
  }

  &::after {
    border-#{$borderSide}-color: $cdr-color-background-primary;
    border-width: 9px;
  }
}

.cdr-popover{
  padding: $cdr-space-inset-one-x-squish;
  background: $cdr-color-background-primary;
  border: 1px solid $cdr-color-border-secondary;
  box-shadow: $cdr-prominence-elevated;
  min-width: 286px;
  position: absolute;
  z-index: 100;

  &::before,
  &::after {
    content: '';
    position: absolute;
    margin: auto;
    width: 0;
    height: 0;
  }

  // Up and down arrows
  &__arrow--up,
  &__arrow--down {
    &,
    &::before,
    &::after {
      left: 50%;
      transform: translateX(-50%);
    }
  }

  &__arrow--up {
    @include arrow(bottom);
    top: calc(100% + 2rem);
  }

  &__arrow--down {
    @include arrow(top);
    bottom: calc(100% + 2rem);
  }


  // Left and right arrows
  &__arrow--left,
  &__arrow--right {
    &,
    &::before,
    &::after {
      top: 50%;
      transform: translateY(-50%);
    }
  }

  &__arrow--left {
    @include arrow(right);
    left: calc(100% + 2rem);
  }

  &__arrow--right {
    @include arrow(left);
    right: calc(100% + 2rem);
  }
}
