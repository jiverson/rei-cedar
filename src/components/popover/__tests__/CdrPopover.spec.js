import { shallowMount } from '@vue/test-utils';
import CdrPopover from 'componentdir/popover/CdrPopover';

describe('CdrPopover', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(CdrPopover);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('is not visible by default', () => {
    const wrapper = shallowMount(CdrPopover);
    expect(wrapper.html()).toBe('');
  });

  it('applies correct arrow class', () => {
    const wrapper = shallowMount(CdrPopover, {
      propsData: {
        position: 'left',
        opened: true
      },
    });
    expect(wrapper.classes()).toContain('cdr-popover__wrapper--left');
  });

  it('applies tooltip class if trigger prop is set', () => {
    const wrapper = shallowMount(CdrPopover, {
      propsData: {
        position: 'left',
        opened: true,
        trigger: 'tooltip'
      },
    });
    expect(wrapper.classes()).toContain('cdr-popover--tooltip');
  });
});
