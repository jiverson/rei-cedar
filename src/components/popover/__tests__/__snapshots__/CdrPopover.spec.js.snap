// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrPopover matches snapshot 1`] = `
<div
  class="cdr-popover--wrapper"
>
  <div />
  <cdr-popup-stub
    autoposition="true"
    position="up"
    role="dialog"
  >
    <div
      class="cdr-popover__container"
    >
      <div
        class="cdr-popover__content"
      />
      <cdr-button-stub
        aria-label="Close"
        class="cdr-popover__close-button"
        icononly="true"
        size="small"
        space=""
        tag="button"
        type="button"
      >
        <icon-x-sm-stub
          inheritcolor="true"
          size="medium"
          space=""
        />
      </cdr-button-stub>
    </div>
  </cdr-popup-stub>
</div>
`;
